Library/Homebrew/command.rb

```diff
@@ -0,0 +1,157 @@
+module Homebrew
+  class Command
+    attr_reader :command_name, :valid_options, :description, :help_output,
+      :man_output
+
+    def initialize_variables
+      @valid_options = []
+      @root_options = []
+      @argv = ARGV.dup
+    end
+
+    def options(&block)
+      initialize_variables
+      @parent = nil
+      instance_eval(&block)
+      generate_help_and_manpage_output
+    end
+
+    def cmd_name(cmd)
+      @command_name = cmd
+    end
+
+    def add_valid_option(option_hash)
+      @valid_options.push(option_hash)
+    end
+
+    def option(*args, **option_hash, &block)
+      option_hash[:option] = "--#{args[0]}"
```

Comment:
>
```
      if @parent.nil?
        @root_options.push(option_name)
      else
        hash = @valid_options.find { |x| x[:option] == @parent }
        if hash[:child_options].nil?
          hash[:child_options] = [option_name]
        else
          hash[:child_options].push(option_name)
        end
      end
```
Either an `option` has some parent option or it does not. If it has a parent option (i.e. if it is present inside the `block` of some other option), then, it's name is included in the `:child_options ` field of the parent's option hash present in `@valid_options`

> 
```
      add_valid_option(option_hash)
```
Inserts the option information in the `@valid_options`

> 
```
      return unless block_given?
      old_parent = @parent
      @parent = option_name
      instance_eval(&block)
      @parent = old_parent
```
Evaluates the block (nest) of an option